### PR TITLE
[common-go] Add keyed gRPC rate limits

### PR DESCRIPTION
**Pull request by csweichel**
## Description
This PR adds support for message-keyed gRPC rate limits to common-go.
Using this functionality, we can e.g. add per-user rate limits in ws-manager.

For example:
```JSON
{
    "rateLimits": {
        "/wsman.WorkspaceManager/StartWorkspace": {
            "bucketSize": 5,
            "refillInterval": "1s",
            "key": "metadata.owner",
            "keyCacheSize": 200
        }
    }
}
```

would impose a 5 workspaces per second rate limit per workspace owner.

## How to test
- See unit test
- Manually add the config to ws-manager and use loadgen

## Release Notes

```release-note
[ws-manager] Support more fine-grained rate limits
```

### CODE REVIEW

File: test/go.sum [hunk 1]
```diff
@@ -1429,15 +1426,16 @@ k8s.io/kubectl v0.23.5/go.mod h1:lLgw7cVY8xbd7o637vOXPca/w6HC205KsPCRDYRCxwE=
...
 k8s.io/utils v0.0.0-20210802155522-efc7438f0176/go.mod h1:jPW/WVKK9YHAvNhRxK0md/EJ228hCsBRufyofKtW8HA=
 k8s.io/utils v0.0.0-20211116205334-6203023598ed h1:ck1fRPWPJWsMd8ZRFsWc6mh/zHp5fZ/shhbrgPUxDAE=
 k8s.io/utils v0.0.0-20211116205334-6203023598ed/go.mod h1:jPW/WVKK9YHAvNhRxK0md/EJ228hCsBRufyofKtW8HA=
 rsc.io/binaryregexp v0.2.0/go.mod h1:qTv7/COck+e2FymRvadv62gMdZztPaShugOCi3I+8D8=
 rsc.io/quote/v3 v3.1.0/go.mod h1:yEA65RcK8LyAZtP9Kv3t0HmxON59tX3rD+tICJqUlj0=
 rsc.io/sampler v1.3.0/go.mod h1:T1hPZKmBbMNahiBKFy5HrXp6adAjACjK9JXDnKaTXpA=
 sigs.k8s.io/apiserver-network-proxy/konnectivity-client v0.0.30/go.mod h1:fEO7lRTdivWO2qYVCVG7dEADOMo/MLDCVr8So2g88Uw=
-sigs.k8s.io/controller-runtime v0.11.2 h1:H5GTxQl0Mc9UjRJhORusqfJCIjBO8UtUxGggCwL1rLA=
-sigs.k8s.io/controller-runtime v0.11.2/go.mod h1:P6QCzrEjLaZGqHsfd+os7JQ+WFZhvB8MRFsn4dWF7O4=
+sigs.k8s.io/controller-runtime v0.9.0 h1:ZIZ/dtpboPSbZYY7uUz2OzrkaBTOThx2yekLtpGB+zY=
```
Comment (aledbf) [inline test/go.sum:1437]: Please revert this change

File: test/go.mod [hunk 2]
```diff
@@ -139,10 +140,6 @@ require (
 	sigs.k8s.io/yaml v1.3.0 // indirect
 )
 
-replace sigs.k8s.io/controller-runtime => sigs.k8s.io/controller-runtime v0.11.0
```
Comment (aledbf) [inline test/go.mod:142]: Please revert this change

Comment (csweichel) [reply test/go.mod:142]: Which one is correct?
```

replace sigs.k8s.io/controller-runtime => sigs.k8s.io/controller-runtime v0.11.0

replace sigs.k8s.io/controller-runtime => sigs.k8s.io/controller-runtime v0.11.2
```

Comment (aledbf) [reply test/go.mod:142]: v0.11.2

**Comment by easyCZ** [comment]
https://github.com/grpc-ecosystem/go-grpc-middleware offers a rate limit implementation, this may be possibly cheaper to use than building our own

**Comment by csweichel** [comment]
> https://github.com/grpc-ecosystem/go-grpc-middleware offers a rate limit implementation, this may be possibly cheaper to use than building our own.
> 
> I've used it in the past and worked well. It has built in support for sharing the rate limits across instances (but can be used without it).

We looked at that implementation prior to rolling our own.
AFAIK it does not support
- per method rate limiting
- "keyed" rate limiting as is introduced in this PR.

**Review by easyCZ — COMMENTED** [review]
Looking good, left a couple comments. Could you please also add metrics to expose the following:
1. Rate limited count, by method
2. Hit rate on the LRU

Can be in seperate follow-up PRs, but we should be able to observe this before we deploy it to prod.

File: components/common-go/grpc/ratelimit.go [hunk 3]
```diff
@@ -6,60 +6,143 @@ package grpc
...
+		var limit *rate.Limiter
+		if f.Key == nil {
+			limit = f.GlobalLimit
+		} else {
+			key, err := f.Key(req)
+			if err != nil {
+				return nil, err
+			}
+
+			f.KeyedLimit.ContainsOrAdd(key, f.RateLimit.Limiter())
```
Comment (geropl) [inline components/common-go/grpc/ratelimit.go:132]: question: this immediately throws creates and trashes a limiter struct on stack AFAIU which could be avoided by not using `ContainsOrAdd`.

Comment (csweichel) [reply components/common-go/grpc/ratelimit.go:132]: Agreed - not doing it this way requires extra locking because this operation can happen concurrently. 

We could also ignore the concurrency for the time being and go with a `Peek` followed by an `Add`, running this risk to mis-account for a user during very high call-rate situations.

**Review by geropl — APPROVED** [review]
Code LGTM, and seems to work. Left minor comments.

File: components/common-go/grpc/ratelimit.go [hunk 4]
```diff
@@ -6,60 +6,143 @@ package grpc
...
+	RateLimit RateLimit
+
+	GlobalLimit *rate.Limiter
+	Key         keyFunc
+	KeyedLimit  *lru.Cache
 }
 
 // UnaryInterceptor creates a unary interceptor that implements the rate limiting
 func (r RatelimitingInterceptor) UnaryInterceptor() grpc.UnaryServerInterceptor {
 	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (resp interface{}, err error) {
```
[the remark this answers is not in the document]
Comment (csweichel) [reply components/common-go/grpc/ratelimit.go:117]: they indicate what's returned though :)

**Comment by csweichel** [comment]
Added metrics

File: components/common-go/grpc/ratelimit.go [hunk 5]
```diff
@@ -6,60 +6,201 @@ package grpc
...
+}
+
+func (r RateLimit) Limiter() *rate.Limiter {
+	return rate.NewLimiter(rate.Every(time.Duration(r.RefillInterval)), int(r.BucketSize))
 }
 
 // NewRatelimitingInterceptor creates a new rate limiting interceptor
 func NewRatelimitingInterceptor(f map[string]RateLimit) RatelimitingInterceptor {
-	funcs := make(map[string]ratelimitedFunction, len(f))
+	callCounter := prometheus.NewCounterVec(prometheus.CounterOpts{
```
Comment (easyCZ) [inline components/common-go/grpc/ratelimit.go:42]: I believe the metrics here also need to be registered, to be picked up by prometheus, otherwise they are just local counters which won't show up on `/metrics` endpoint.

See https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#Register

Comment (easyCZ) [reply components/common-go/grpc/ratelimit.go:42]: In this case, I'd actually hoist the counter definition outside of this instantiation and make it package private variable. You can use https://pkg.go.dev/github.com/prometheus/client_golang/prometheus/promauto to simplify the registration.

**Review by sagor999 — APPROVED** [review]
just need to fix metrics name copy paste error.

File: components/common-go/grpc/ratelimit.go (hunk 5 above)
Comment (easyCZ) [reply components/common-go/grpc/ratelimit.go:42]: My bad, missed this. Normally, I'd expect the registration to happen in this file and automatically without requiring an extra call. The risk here is that a component uses the rate limiting interceptor but doesn't register the metric.

The other issue is that this would panic on startup, but CI doesn't catch this.

Comment (csweichel) [reply components/common-go/grpc/ratelimit.go:42]: For a while now we've adopted the pattern to make things implement the `prometheus.Collector` interface so that they can be registered.
Testing is just as easy as passing in the registry, but it doesn't mandate that someone uses metrics.
The "registration error at runtime" behaviour is also the same.

**Comment by csweichel** [comment]
fixed the metrics name

File: components/common-go/grpc/ratelimit.go (hunk 5 above)
Comment (easyCZ) [reply components/common-go/grpc/ratelimit.go:42]: Thanks. I'd personally like the need for registration to be more explicit (and to use DI to pass the registry in) but given this is an existing pattern, better to stick to it than introduce yet another. Thanks for the info.
